cluster_experiments/perturbator.py

```diff
@@ -350,6 +351,274 @@ def _sample_beta_effect(
+        target_col (str, optional): name of the target_col to use as the outcome. Defaults to "target".
+        treatment_col (str, optional): the name of the column that contains the treatment. Defaults to "treatment".
+        treatment (str, optional): name of the treatment to use as the treated group. Defaults to "B".
+        scale (Optional[float], optional): the scale of the effect distribution. Defaults to None.
+        range_min (float, optional): the minimum value of the target range. Defaults to -0.8.
+        range_max (float, optional): the maximum value of the target range. Defaults to 5.
+        cluster_cols (Optional[List[str]], optional): the columns to use for clustering. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        average_effect: Optional[float] = None,
+        target_col: str = "target",
+        treatment_col: str = "treatment",
+        treatment: str = "B",
+        scale: Optional[float] = None,
+        range_min: float = -0.8,
+        range_max: float = 5,
+        cluster_cols: Optional[List[str]] = None,
+    ):
+        super().__init__(average_effect, target_col, treatment_col, treatment, scale)
+        self._range_min = range_min
+        self._range_max = range_max
+        self._cluster_cols = cluster_cols
+        self.cluster_col = None
+
+    def get_cluster_perturbator(
+        self, df: pd.DataFrame, average_effect: Optional[float] = None
+    ) -> Perturbator:
+        average_effect = self.get_average_effect(average_effect)
+        self.check_beta_positive_effect(df, average_effect)
+        scale = self.get_scale(average_effect)
+        sampled_effect = self._sample_beta_effect(average_effect, scale, 1)
+        cluster_perturbator = BetaRelativePerturbator(
+            average_effect=sampled_effect,
+            target_col=self.target_col,
+            treatment_col=self.treatment_col,
+            treatment=self.treatment,
+            range_min=self._range_min,
+            range_max=self._range_max,
+        )
+        return cluster_perturbator
+
+    def _set_cluster_col(
+        self, df: pd.DataFrame, cluster_cols: Optional[List[str]] = None
+    ):
+        cluster_cols = cluster_cols or self._cluster_cols
+        if not cluster_cols:
+            raise ValueError("cluster_cols must be set!")
+
+        if not isinstance(cluster_cols, list):
+            raise ValueError(
+                f"cluster_cols must be of type List[str], got type {type(cluster_cols)}"
+            )
+
+        self.cluster_col = "_cluster_" + "_".join(cluster_cols)
+        if self.cluster_col in df.columns:
+            raise ValueError(
+                f"Cannot use self.cluster_col={self.cluster_col} as clustering column, as it already exists in the input dataframe!"
+            )
+
+        df = df.copy()
+        cluster = pd.Series("", index=df.index)
+        for col in cluster_cols:
```

Comment:
can't this be done with a sum? like this: https://github.com/david26694/cluster-experiments/blob/main/cluster_experiments/experiment_analysis.py#L46

Reply:
So much nicer! Thanks!